### PR TITLE
feat(ivy): i18n compiler - i18nStart and i18nEnd support

### PR DESCRIPTION
**Pull request by AndrewKushnir**
## PR Description
This PR introduces `i18nStart` and `i18nEnd` instructions support in i18n compiler, according to the [I18n Design Doc](https://github.com/angular/angular/blob/dfbb79ac72c1715a1506d2ab8441ef036a5f7bca/packages/core/src/render3/i18n.md). The main idea is to maintain an instance of `I18nContext` between start and end instructions, which keeps track of all i18n-related aspects (accumulates content, bindings, etc). When we enter a nested template (for ex. as a result of *ngIf), the top-level context is being passed down to the nested component, which uses this context to generate a child instance of `I18nContext` class (to handle nested template) and at the end, reconciles it back with the parent context.

### Notes

- `<ng-template i18n>`, `<ng-content i18n>` and `<ng-container i18n>` are not supported yet (coming next)
- the logic of `goog.getMsg` generation remained as is (was not changed in this PR)
- no ICU support introduced in this PR (coming next)

### Next steps (in followup PRs)

- `<ng-*>` tags support
- updates to tests logic to support � symbol in templates
- (to be discussed) `ConstantPool` refactoring to provide API to support Translation needs. Currently (historically) Translation-related logic to generate  constants, is located in `ConstantPool` - we can probably extract it out to `i18n.ts`.
- (to be discussed) introduce `i18n` instruction in case there is no other content in between `i18nStart` and `i18nEnd`, a-la `element` instruction

## PR Type

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by ocombe — CHANGES_REQUESTED** [review]

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 1]
```diff
@@ -310,24 +388,441 @@ describe('i18n support in the view compiler', () => {
...
+          `
         }
       };
 
-      const template = `
-      const $r1$ = {"b":[2], "i":[4, 6]};
-      const $r2$ = {"i":[13]};
-    `;
+      const template = String.raw `
+        const $MSG_APP_SPEC_TS_0$ = goog.getMsg("\uFFFD#0\uFFFDMy i18n block #1\uFFFD/#0\uFFFD");
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: Good catch, thanks! I've updated the code and the necessary tests to handle that.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 2]
```diff
@@ -310,24 +388,441 @@ describe('i18n support in the view compiler', () => {
...
+            export class MyComponent {}
+
+            @NgModule({declarations: [MyComponent]})
+            export class MyModule {}
+          `
+        }
+      };
+
+      const template = String.raw `
+        const $MSG_APP_SPEC_TS_0$ = goog.getMsg("\uFFFD#0\uFFFD My i18n block #\uFFFD0\uFFFD \uFFFD#2\uFFFDPlain text in nested element\uFFFD/#2\uFFFD\uFFFD/#0\uFFFD");
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: Agree, I've added the code to trim the content while assembling i18n payload.

**Comment by AndrewKushnir** [comment]
Thanks for the review @ocombe! 👍
I've updated the code based on your comments.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 3]
```diff
@@ -379,6 +379,41 @@ describe('i18n support in the view compiler', () => {
   });
 
   describe('nested nodes', () => {
+    it('should not produce instructions for empty content', () => {
```
Comment (ocombe) [inline packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: I'm not sure about this one, is it possible to have an empty english translation that would not be empty in another language?

File: packages/compiler/src/render3/view/i18n.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  let acc = '';
+  const lastIdx = strings.length - 1;
+  for (let i = 0; i < lastIdx; i++) {
+    acc += `${strings[i]}${wrapI18nPlaceholder(bindingStartIndex + i, contextId)}`;
+  }
+  acc += strings[lastIdx];
+  return acc;
+}
+
+function getSeqNubmerGenerator(startsAt: number = 0): () => number {
```
Comment (JoostK) [inline packages/compiler/src/render3/view/i18n.ts]: Nubmer -> Nu**mb**er

File: packages/compiler/src/constant_pool.ts [hunk 5]
```diff
@@ -113,6 +114,30 @@ export class ConstantPool {
     return fixup;
   }
 
+  getDeferredTranslationConst(suffix: string): o.ReadVarExpr {
+    const index = this.statements.push(new o.ExpressionStatement(o.NULL_EXPR)) - 1;
```
Comment (JoostK) [inline packages/compiler/src/constant_pool.ts]: Could you rephrase the `get` part in this method's name, as it does have side-effects.

File: packages/compiler/src/render3/view/i18n.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  let id: string|undefined;
+
+  if (i18n) {
+    // TODO(vicb): figure out how to force a message ID with closure ?
+    const idIndex = i18n.indexOf(I18N_ID_SEPARATOR);
+    const descIndex = i18n.indexOf(I18N_MEANING_SEPARATOR);
+    let meaningAndDesc: string;
+    [meaningAndDesc, id] =
+        (idIndex > -1) ? [i18n.slice(0, idIndex), i18n.slice(idIndex + 2)] : [i18n, ''];
+    [meaning, description] = (descIndex > -1) ?
```
Comment (JoostK) [inline packages/compiler/src/render3/view/i18n.ts]: This logic might not be yours (per the `TODO(vicb)`) but I think it's strange that the `*Index` vars are computed on the initial string in full, then `descIndex` is used on a substring here.

File: packages/compiler/src/render3/view/i18n.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  return () => current++;
+}
+export class I18nContext {
+  private id: number;
+  private content: string = '';
+  private bindings = new Set<o.Expression>();
+
+  constructor(
+      private index: number, private templateIndex: number|null = null, private ref: any,
+      private level: number = 0, private uniqueIdGen?: (() => number)) {
```
Comment (JoostK) [inline packages/compiler/src/render3/view/i18n.ts]: nit: the parenthesis around the `() => number` should't be necessary

File: packages/compiler/src/render3/view/i18n.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+  let current = startsAt;
+  return () => current++;
+}
+export class I18nContext {
+  private id: number;
+  private content: string = '';
+  private bindings = new Set<o.Expression>();
+
+  constructor(
+      private index: number, private templateIndex: number|null = null, private ref: any,
```
Comment (JoostK) [inline packages/compiler/src/render3/view/i18n.ts]: The `= null` can't be here with the remaining parameters not having a default value.

**Review by JoostK — COMMENTED** [review]
Compliments on your PR descriptions, it's really appreciated! 🥇

**Comment by AndrewKushnir** [comment]
@JoostK thanks for your comments! 👍 I've performed the necessary updates.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: My idea was to make sure we are treating empty content properly: the compiler code doesn't fail and the output has the right shape.

**Review by ocombe — APPROVED** [review]

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 9]
```diff
@@ -310,24 +398,476 @@ describe('i18n support in the view compiler', () => {
...
+
+      const template = String.raw `
+        const $MSG_APP_SPEC_TS_0$ = goog.getMsg("My i18n block #\uFFFD0\uFFFD\uFFFD#2\uFFFDPlain text in nested element\uFFFD/#2\uFFFD");
+        const $MSG_APP_SPEC_TS_1$ = goog.getMsg("My i18n block #\uFFFD0\uFFFD\uFFFD#6\uFFFD\uFFFD#7\uFFFD\uFFFD#8\uFFFDMore bindings in more nested element: \uFFFD1\uFFFD\uFFFD/#8\uFFFD\uFFFD/#7\uFFFD\uFFFD/#6\uFFFD");
+        …
+        template: function MyComponent_Template(rf, ctx) {
+          if (rf & 1) {
+            $r3$.ɵelementStart(0, "div");
+            $r3$.ɵi18nStart(1, $MSG_APP_SPEC_TS_0$);
+            $r3$.ɵelementStart(2, "span");
```
Comment (mhevery) [inline packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]:
```suggestion
            $r3$.ɵelement(2, "span");
```

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 10]
```diff
@@ -310,24 +398,476 @@ describe('i18n support in the view compiler', () => {
...
+            $r3$.ɵelementEnd();
```
Comment (mhevery) [inline packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]:
```suggestion
```

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts (hunk 9 above)
Comment (mhevery) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: also consider `i18n(..)` which merges `start/end`

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 11]
```diff
@@ -43,32 +39,37 @@ describe('i18n support in the view compiler', () => {
       };
 
       const template = `
-        const $msg_1$ = goog.getMsg("Hello world");
-        const $msg_2$ = goog.getMsg("farewell");
+        /**
+         * @desc descA
```
Comment (mhevery) [inline packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]:
```suggestion
         * @desc [BACKUP_MESSAGE_ID:idA] descA
```

File: packages/compiler/src/render3/view/i18n.ts [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+  return () => current++;
+}
+
+export type I18nMeta = {
+  id?: string,
+  description?: string,
+  meaning?: string
+};
+
+export class I18nContext {
```
Comment (mhevery) [inline packages/compiler/src/render3/view/i18n.ts]: Add comment which explains:
- what it do.
- how does it do it.
- Why was this chosen.

Comment (mhevery) [reply packages/compiler/src/render3/view/i18n.ts]: Add unit test for this class directly.

**Review by mhevery — APPROVED** [review]
